src/main/java/fi/aalto/cs/apluscourses/presentation/base/SelectableNodeViewModel.java

```diff
@@ -35,18 +35,19 @@ public SelectableNodeViewModel(@NotNull T model,
    * @param filter A filter.
    * @return True, if the filter applies to this node or one of its descendants, otherwise false.
    */
-  public boolean applyFilter(Filter filter) {
-    if (Thread.currentThread().isInterrupted()) {
-      return true;
+  public Optional<Boolean> applyFilter(Filter filter) {
+    Optional<Boolean> result = filter.apply(this);
+    for (SelectableNodeViewModel<?> child : children) {
+      if (Thread.currentThread().isInterrupted()) {
+        return Optional.empty();
+      }
+      Optional<Boolean> childResult = child.applyFilter(filter);
+      if (childResult.isPresent()) {
+        result = Optional.of(result.orElse(false) || childResult.get());
+      }
```

Comment:
Am I understanding the logic here correctly, that if this node does NOT pass the filter (`false`) but one of the children does pass the filter (`true`), then this node becomes visible after all?

Reply:
You were right. I changed it now so that the descendants matter only if `filter.apply(this)` returns empty.